machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+	MachineConfigSelector *metav1.LabelSelector `json:"machineConfigSelector,omitempty"`
+
+	// Node selector for Machines
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// If true, changes to this machine config pool should be stopped.
+	// This includes generating new desiredMachineConfig and update of machines.
+	Paused bool `json:"paused"`
+
+	// MaxUnavailable specifies the percentage or constant number of machines that can be updating at any given time.
+	// default is 1.
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable"`
+}
+
+// MachineConfigPoolStatus is the status for MachineConfigPool resource.
+type MachineConfigPoolStatus struct {
+	// The generation observed by the controller.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// The current MachineConfig object for the machine config pool.
+	Configuration MachineConfigPoolStatusConfiguration `json:"configuration"`
+
+	// Total number of machines in the machine config pool.
+	MachineCount int32 `json:"machineCount"`
+
+	// Total number of machines targeted by the pool that have the CurrentMachineConfig as their config.
+	UpdatedMachineCount int32 `json:"updatedMachineCount"`
+
+	// Total number of ready machines targeted by the pool.
+	ReadyMachineCount int32 `json:"readyMachineCount"`
+
+	// Total number of unavailable (non-ready) machines targeted by the pool.
+	// A node is marked unavailable if it is in updating state or NodeReady condition is false.
+	UnavailableMachineCount int32 `json:"unavailableMachineCount"`
+
+	// Total number of machines marked degraded (or unreconcilable).
+	// A node is marked degraded if applying a configuration failed..
+	DegradedMachineCount int32 `json:"degradedMachineCount"`
+
+	// Represents the latest available observations of current state.
+	Conditions []MachineConfigPoolCondition `json:"conditions"`
+}
+
+// MachineConfigPoolStatusConfiguration stores the current configuration for the pool, and
+// optionally also stores the list of machineconfig objects used to generate the configuration.
+type MachineConfigPoolStatusConfiguration struct {
+	corev1.ObjectReference `json:",inline"`
+
+	// source is the list of machineconfigs that were used to generate the single machineconfig object specified in `content`.
+	// +optional
+	Source []corev1.ObjectReference `json:"source,omitempty"`
+}
+
+// MachineConfigPoolCondition contains condition information for an MachineConfigPool.
+type MachineConfigPoolCondition struct {
+	// Type of the condition, currently ('Done', 'Updating', 'Failed').
+	Type MachineConfigPoolConditionType `json:"type"`
+
+	// Status of the condition, one of ('True', 'False', 'Unknown').
+	Status corev1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the timestamp corresponding to the last status
+	// change of this condition.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
```

Comment:
the comment should use the json name of the field because they show up in kubectl explain and docs.

And everywhere else.